cpp/src/transpose/transpose.cu

```diff
@@ -0,0 +1,172 @@
+/* Copyright 2018 NVIDIA Corporation.  All rights reserved. */
+
+#include <cudf.h>
+#include <stdio.h>
+#include <thrust/for_each.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <cub/cub.cuh>
+#include <memory>
+#include "dataframe/cudf_table.cuh"
+#include "utilities/nvtx/nvtx_utils.h"
+#include "utilities/type_dispatcher.hpp"
+#include <chrono>
+
+constexpr int WARP_SIZE = 32;
+constexpr int MAX_GRID_SIZE = (1<<16)-1;
+
+template <typename ColumnType>
+__global__
+void gpu_transpose(ColumnType **in_cols, ColumnType **out_cols,
+                  gdf_size_type ncols, gdf_size_type nrows)
+{
+  gdf_size_type x = blockIdx.x * blockDim.x + threadIdx.x;
+  gdf_size_type y = blockIdx.y * blockDim.y + threadIdx.y;
+    
+  gdf_size_type stride_x = blockDim.x * gridDim.x;
+  gdf_size_type stride_y = blockDim.y * gridDim.y;
+
+  for(gdf_size_type i = x; i < ncols; i += stride_x)
+  {
+    for(gdf_size_type j = y; j < nrows; j += stride_y)
+    {
+      out_cols[j][i] = in_cols[i][j];
+    }
+  }
+}
+
+__global__
+void gpu_transpose_valids(gdf_valid_type **in_cols_valid,
+                          gdf_valid_type **out_cols_valid,
+                          gdf_size_type ncols, gdf_size_type nrows)
+{
+  using MaskType = uint32_t;
+  constexpr uint32_t BITS_PER_MASK{sizeof(MaskType) * 8};
+
+  gdf_size_type x = blockIdx.x * blockDim.x + threadIdx.x;
+  gdf_size_type y = blockIdx.y * blockDim.y + threadIdx.y;
+
+  gdf_size_type stride_x = blockDim.x * gridDim.x;
+  gdf_size_type stride_y = blockDim.y * gridDim.y;
+
+  MaskType* const __restrict__ out_mask32 =
+    reinterpret_cast<MaskType*>(out_cols_valid[y]);
+
+  for(gdf_size_type i = x; i < ncols; i += stride_x)
+  {
+    for(gdf_size_type j = y; j < nrows; j += stride_y)
+    {
+      auto active_threads = __ballot_sync(0xffffffff, x < ncols && y < nrows);
```

Comment:
Great catch. I didn't catch this in testing because all the threads at the end of the warp which shouldn't be active correspond to the area in the valid mask that's just padding at the end. And the test didn't check whether those bits were 0s or 1s. BTW, is it a requirement that the padding bits be 0s?

Reply:
> BTW, is it a requirement that the padding bits be 0s?

Bits outside of `[0, num_rows)` are undefined and if anyone relies on them being `0`, that's broken code.